debug/org.eclipse.debug.core/core/org/eclipse/debug/core/DebugPlugin.java

```diff
@@ -727,22 +735,43 @@ public void stop(BundleContext context) throws Exception {
 	public void start(BundleContext context) throws Exception {
 		super.start(context);
 		new DebugOptions(context);
-		ResourcesPlugin.getWorkspace().addSaveParticipant(getUniqueIdentifier(),
-				new ISaveParticipant() {
-					@Override
-					public void saving(ISaveContext saveContext) throws CoreException {
-						if (fExpressionManager != null) {
-							fExpressionManager.storeWatchExpressions();
+
+		fWorkspaceServiceTracker = new ServiceTracker<>(context, IWorkspace.class, null) {
```

Comment:
Instead of extending ServiceTracker you should use a `ServiceTrackerCustomizer`, that can the return a custom object (e.g. a `Runnable` that performs the cleanup in this case seems useful), then this will even work if there are multiple workspaces (even if unlikley), or the workspace plugin shut down before your plugin (what can happen as shutdown might be performed in reverse order), that way you can simply close the tracker to perform the removal

Reply:
> A ServiceTracker implements already a ServiceTrackerCustomizer

because it is used internally if you don't pass your own in the constructor but is only meant for tracking simple services where no further processing is required.

> and by that allows to only overwrite the methods which are relevant

It does not hurt to leave one method (e.g. update) empty implemented, that is usually much more readable and flexible than extending the tracker.

> What I would rather do is to also do the cleanup in the ServiceTracker (instead of the stop method)

this is what I would recommend here, as otherwhise there is now a way that the stop method return null...

> Returning a Runnable seems overly complex because the ServiceTracker is not supposed to be used with getService() nor does it need to keep track of the bound SaveParticipant.

I won't assume `Runnable` complex at the end it does only matter that cleanup is done when the service about to disappear (or the tracker is closed), not only when the bundle is stopped.

Reply:
I now did the cleanup in the servicetrackercustomizer as well. That way it should be called when the tracker is closed or the service disappears. \

> It does not hurt to leave one method (e.g. update) empty implemented, that is usually much more readable and flexible than extending the tracker.

I don't agree with that statement, overriding ServiceTracker also nicely encapsulates dealing with the bundle context and the anonymous inner class should be pretty much straight-forward here.